app/response_processor.py

```diff
@@ -9,130 +8,101 @@
-
-        endpoint = receipt.get_receipt_endpoint(decrypted_json)
-        if endpoint is None:
-            return False
+        if "metadata" not in decrypted:
+            raise BadMessageError("Missing metadata")
+        return
 
-        xml = receipt.get_receipt_xml(decrypted_json)
+    def _encode(self, decrypted):
+        xml = receipt.get_receipt_xml(decrypted)
         if xml is None:
-            return False
+            raise BadMessageError("Unable to generate xml from message")
+        return xml
 
-        headers = receipt.get_receipt_headers()
+    def _send_receipt(self, decrypted, xml):
+        endpoint = receipt.get_receipt_endpoint(decrypted)
+        if endpoint is None:
+            raise BadMessageError("Unable to determine delivery endpoint from message")
 
-        response = self.remote_call(
-            endpoint,
-            data=xml.encode("utf-8"),
-            headers=headers,
-            verify=False,
-            auth=(settings.RECEIPT_USER, settings.RECEIPT_PASS))
-        return self.response_ok(response)
+        headers = receipt.get_receipt_headers()
+        auth = (settings.RECEIPT_USER, settings.RECEIPT_PASS)
 
-    def remote_call(self, request_url, json=None, data=None, headers=None, verify=True, auth=None):
         try:
-            self.logger.info("Calling service", request_url=request_url)
-            r = None
-
-            if data:
-                r = session.post(request_url, data=data, headers=headers, verify=verify, auth=auth)
-            else:
-                r = session.get(request_url, headers=headers, verify=verify, auth=auth)
+            self.logger.info("Calling service", request_url=endpoint)
+            res = session.post(endpoint, data=xml, headers=headers, verify=False, auth=auth)
 
-            return r
+            if res.status_code == 400:
+                self.logger.error("Receipt rejected by endpoint", request_url=res.url, status_code=400)
+                raise BadMessageError("Failure to send receipt")
 
-        except MaxRetryError:
-            self.logger.error("Max retries exceeded (5)", request_url=request_url)
+            elif res.status_code != 200 and res.status_code != 201:
+                # Endpoint may be temporarily down
+                self.logger.error("Bad response from endpoint", request_url=res.url, status_code=res.status_code)
+                raise RetryableError("Failure to send receipt")
 
-    def response_ok(self, res):
-        if res.status_code == 200 or res.status_code == 201:
-            self.logger.info("Returned from service", request_url=res.url, status_code=res.status_code)
-            return True
+            else:
+                self.logger.info("Returned from service", request_url=res.url, status_code=res.status_code)
+                return
 
-        else:
-            self.logger.error("Returned from service", request_url=res.url, status_code=res.status_code)
-            return False
+        except MaxRetryError:
```

Comment:
Yes - but different retries. This isn't the queue - it's the number of times the http client attempts top connect to the endpoint before giving up.